File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionXml.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;
import co.nlighten.jsontransform.formats.xml.XmlFormat;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;

/*
 * For tests
 * @see TransformerFunctionXmlTest
 */
public class TransformerFunctionXml extends TransformerFunction {
    static final Logger logger = LoggerFactory.getLogger(TransformerFunctionXml.class);

    public TransformerFunctionXml() {
        super(FunctionDescription.of(
                Map.of(
                        "root", ArgumentType.of(ArgType.String).position(0).defaultIsNull(true),
                        "xslt", ArgumentType.of(ArgType.String).position(1).defaultIsNull(true),
                        "indent", ArgumentType.of(ArgType.Boolean).position(2).defaultBoolean(false)
                )
        ));
    }

    @Override
    public Object apply(FunctionContext context) {
        var obj = context.getJsonElement(null);
        if (obj == null)
            return null;
        try {
            var rootName = context.getString("root");
            // TODO: how to create the format once?
            var xml = new XmlFormat(context.getAdapter()).obj2xml(obj, rootName);
            var xslt = context.getString("xslt");
            if (xslt != null && !xslt.isBlank()) {
                var transformer = XmlFormat.createXSLTTransformer(xslt);
                xml = XmlFormat.xmlTransform(xml, transformer);
            }
            if (context.getBoolean("indent")) {
                xml = XmlFormat.indentXml(xml);
            }
            return xml;
        } catch (Exception e) {
            logger.warn(context.getAlias() + " failed", e);
            return null;
        }
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionXml.ts
import xml2js from "xml2js";
import TransformerFunction from "./common/TransformerFunction";
import FunctionContext from "./common/FunctionContext";
import { ArgType } from "./common/ArgType";

const DUMMY_ROOT = "R__O__O__T";
class TransformerFunctionXml extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        root: { type: ArgType.String, position: 0, defaultIsNull: true },
        xslt: { type: ArgType.String, position: 1, defaultIsNull: true },
        indent: { type: ArgType.Boolean, position: 2, defaultBoolean: false },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const obj = await context.getJsonElement(null);
    if (obj == null) return null;
    try {
      const rootName = (await context.getString("root")) ?? DUMMY_ROOT;
      const indent = (await context.getBoolean("indent")) ?? undefined;
      const builder = new xml2js.Builder({
        headless: !indent,
        rootName,
        charkey: "$content",
        renderOpts: { pretty: indent },
        xmldec: {
          version: "1.0",
          encoding: "UTF-8",
          standalone: undefined,
        },
      });
      let xml = builder.buildObject(obj);
      if (rootName === DUMMY_ROOT) {
        xml = xml.slice(DUMMY_ROOT.length + 2, -DUMMY_ROOT.length - 3);
      }
      if (indent) {
        // remove new line after xmldec
        xml = xml.replace("?>\n", "?>") + "\n";
      }
      const xslt = await context.getString("xslt");
      if (xslt != null && xslt !== "") {
        throw new Error("XSLT not supported"); // TODO: add XSLT support
      }
      return xml;
    } catch (e: any) {
      console.warn(context.getAlias() + " failed", e);
      return null;
    }
  }
}

export default TransformerFunctionXml;
